File: weapon.py
import pygame
import math
import random
import constants


class Weapon():
    def __init__(self, image, bullet_image):
        self.angle = 0
        self.original_image = image
        self.flip = False
        self.weapon_angle = 0
        self.image = pygame.transform.rotate(self.original_image, self.angle)
        self.bullet_image = bullet_image
        self.rect = self.image.get_rect()
        self.fired = False
        self.last_shot = pygame.time.get_ticks()


    def update(self, player):
        shot_cooldown = 100
        bullet = None
        self.rect.center = player.rect.center

        pos = pygame.mouse.get_pos()
        x_dist = pos[0] - self.rect.centerx
        y_dist = pos[1] - self.rect.centery
        if x_dist < 0:
            self.flip = True
            self.weapon_angle = math.degrees(-math.atan2(x_dist, y_dist))
        else:
            self.flip = False
            self.weapon_angle = math.degrees(math.atan2(x_dist, y_dist))
        self.angle = math.degrees(math.atan2(x_dist, y_dist))

        # get mouseclick
        if pygame.mouse.get_pressed()[0] and self.fired == False and (pygame.time.get_ticks()) - self.last_shot:
            bullet = Bullet(self.bullet_image, self.rect.centerx + constants.BULLET_OFFSET_X, self.rect.centery + constants.BULLET_OFFSET_Y, self.angle)
            self.fired = True
            self.last_shot = pygame.time.get_ticks()
        if pygame.mouse.get_pressed()[0] == False:
            self.fired = False


        return bullet

    def draw(self, surface):
        self.image = pygame.transform.rotate(self.original_image, self.weapon_angle)
        flipped_image = pygame.transform.flip(self.image, self.flip, False)
        surface.blit(flipped_image, ((self.rect.x + constants.WEAPON_OFFSET_X - int(self.image.get_width() / 2)),
                                     self.rect.centery + constants.WEAPON_OFFSET_Y - int(self.image.get_height() / 2)))


class Bullet(pygame.sprite.Sprite):
    def __init__(self, image, x, y, angle):
        pygame.sprite.Sprite.__init__(self)
        self.original_image = image
        self.angle = angle - 90
        self.image = pygame.transform.rotate(self.original_image, self.angle)
        self.rect = self.image.get_rect()
        self.rect.center = (x, y)

        # shooting directions based on angle
        self.dx = math.cos(math.radians(self.angle)) * constants.BULLET_SPEED
        self.dy = -(math.sin(math.radians(self.angle)) * constants.BULLET_SPEED)

    def update(self, scroll_camera, obstacle_tiles, enemy_list):
        self.rect.x += scroll_camera[0] + self.dx
        self.rect.y += scroll_camera[1] + self.dy

        # check collusion with walls
        for obstacle in obstacle_tiles:
            if obstacle[1].colliderect(self.rect):
                self.kill()


        # damage starts at 0
        damage = 0
        damage_pos = None

        # check if bullet is out range
        if self.rect.right < 0 or self.rect.left > constants.WINDOW_WIDTH or self.rect.bottom < 0 or self.rect.top > constants.WINDOW_HEIGHT:
            self.kill()

        # check bullet hit enemy
        for enemy in enemy_list:
            if enemy.rect.colliderect(self.rect) and enemy.alive:
                damage = constants.SMILE_WEAPON + random.randint(-constants.SMILE_WEAPON_RANDOM, constants.SMILE_WEAPON_RANDOM)
                damage_pos = enemy.rect
                enemy.health -= damage
                enemy.hit = True
                self. kill()
                break
        return damage, damage_pos

    def draw(self, surface):
        surface.blit(self.image, ((self.rect.x - int(self.image.get_width() / 2)),
                                  self.rect.centery - int(self.image.get_height() / 2)))

class Shooter_bullet(pygame.sprite.Sprite):
    def __init__(self, image, x, y, target_x, target_y):
        pygame.sprite.Sprite.__init__(self)
        self.original_image = image
        x_distance = target_x - x
        y_distance = - (target_y - y)
        self.angle = math.degrees(math.atan2(x_distance, y_distance))
        self.image = pygame.transform.rotate(self.original_image, self.angle)
        self.rect = self.image.get_rect()
        self.rect.center = (x, y)

        # shooting directions based on angle
        self.dx = math.sin(math.radians(self.angle)) * (constants.SHOOTER_BULLET_SPEED + random.randint(-constants.BULLET_SPEED_RANDOM, constants.BULLET_SPEED_RANDOM))
        self.dy = - (math.cos(math.radians(self.angle)) * (constants.SHOOTER_BULLET_SPEED + random.randint(-constants.BULLET_SPEED_RANDOM, constants.BULLET_SPEED_RANDOM)))

    def update(self, scroll_camera, obstacle_tiles, player):
        self.rect.x += scroll_camera[0] + self.dx
        self.rect.y += scroll_camera[1] + self.dy


        # check if bullet is out range
        if self.rect.right < 0 or self.rect.left > constants.WINDOW_WIDTH or self.rect.bottom < 0 or self.rect.top > constants.WINDOW_HEIGHT:
            self.kill()

        # check collusion with walls
        for obstacle in obstacle_tiles:
            if obstacle[1].colliderect(self.rect):
                self.kill()


        # check bullet hit player
        if player.rect.colliderect(self.rect) and player.hit == False:
            player.hit = True
            player.last_hit = pygame.time.get_ticks()
            player.health -= 20
            self.kill()

    def draw(self, surface):
        surface.blit(self.image, ((self.rect.x - int(self.image.get_width() / 2)),
                                  self.rect.centery - int(self.image.get_height() / 2)))

File: js/Weapon.js
// Weapon class - handles weapon aiming and shooting
class Weapon {
    constructor(scene, playerSprite) {
        this.scene = scene;
        this.playerSprite = playerSprite;
        
        // Create weapon sprite
        this.sprite = scene.add.sprite(playerSprite.x, playerSprite.y, 'gun');
        this.sprite.setDepth(11);
        this.sprite.setScale(CONFIG.WEAPON_SCALE);
        this.sprite.setOrigin(0.5, 0.5);
        this.lastShot = 0;
        // Python: self.fired = False (prevents holding mouse to spam shoot)
        this.fired = false;
    }
    
    update(time, pointer, playerX, playerY, canShoot = true) {
        // Get world point from pointer
        const worldPoint = this.scene.cameras.main.getWorldPoint(pointer.x, pointer.y);
        const xDist = worldPoint.x - playerX;
        const yDist = worldPoint.y - playerY;
        
        // Python: weapon_angle calculation
        // Python: if x_dist < 0: weapon_angle = math.degrees(-math.atan2(x_dist, y_dist))
        // Python: else: weapon_angle = math.degrees(math.atan2(x_dist, y_dist))
        // Python: self.image = pygame.transform.rotate(self.original_image, self.weapon_angle)
        // Python: flipped_image = pygame.transform.flip(self.image, self.flip, False)
        let weaponAngle;
        let shouldFlip = false;
        
        const angleRad = Math.atan2(xDist, yDist);
        
        // Python: if x_dist < 0: flip=True, weapon_angle=-atan2(x_dist, y_dist)
        // Python: else: flip=False, weapon_angle=atan2(x_dist, y_dist)
        // Python: rotate(original, weapon_angle), then flip(rotated, flip, False)
        // Strictly match Python logic
        if (xDist < 0) {
            // Left side - Python: flip=True, weapon_angle=-atan2
            shouldFlip = true;
            weaponAngle = -angleRad; // Python: -math.atan2(x_dist, y_dist)
        } else {
            // Right side - Python: flip=False, weapon_angle=atan2
            shouldFlip = false;
            weaponAngle = angleRad; // Python: math.atan2(x_dist, y_dist)
        }
        
        // Set weapon position - Python: self.rect.center = player.rect.center
        // Python: weapon rect center = player rect center (they share the same center point)
        // The weapon image center should align with player center
        // In Phaser, setPosition sets sprite center (origin 0.5, 0.5)
        this.sprite.setPosition(
            playerX,
            playerY
        );
        
        // Rotate weapon - Python: pygame.transform.rotate rotates counterclockwise by DEGREES
        // Python: weapon_angle = math.degrees(math.atan2(x_dist, y_dist)) or -math.atan2(x_dist, y_dist)
        // Python: self.image = pygame.transform.rotate(self.original_image, self.weapon_angle)
        // Python: flipped_image = pygame.transform.flip(self.image, self.flip, False)
        // Python's weapon_angle is in DEGREES, Phaser's setRotation uses RADIANS
        // Convert weaponAngle (radians) to degrees, then back to radians for Phaser
        // This ensures the rotation matches Python exactly
        const weaponAngleDeg = Phaser.Math.RadToDeg(weaponAngle);
        const weaponAngleRad = Phaser.Math.DegToRad(weaponAngleDeg);
        
        // Python: rotate first, then flip horizontally
        // Python: self.image = pygame.transform.rotate(self.original_image, self.weapon_angle)
        // Python: flipped_image = pygame.transform.flip(self.image, self.flip, False)
        // pygame.transform.rotate rotates COUNTERCLOCKWISE
        // Phaser's setRotation also rotates COUNTERCLOCKWISE, but rotation direction differs by side
        // Left side: no negation needed (was correct before)
        // Right side: needs negation (was wrong before, now correct)
        // Reset flip state first to avoid compounding effects
        this.sprite.setFlipX(false);
        this.sprite.setFlipY(false);
        
        // Apply rotation (Python rotates by degrees, Phaser by radians)
        // Only negate for right side to match Python's rotation direction
        if (xDist >= 0) {
            // Right side: negate angle
            this.sprite.setRotation(-weaponAngleRad);
        } else {
            // Left side: don't negate (was correct before)
            this.sprite.setRotation(weaponAngleRad);
        }
        
        // Apply horizontal flip if needed (Python: flip=True means horizontal flip)
        // Python: pygame.transform.flip(image, flip, False) - first param is horizontal flip
        // If right side is reversed, try NOT flipping when shouldFlip is false
        this.sprite.setFlipX(shouldFlip);
        
        // Python: self.angle = math.degrees(math.atan2(x_dist, y_dist)) - for bullet direction
        const angle = Math.atan2(xDist, yDist);
        const angleDeg = Phaser.Math.RadToDeg(angle);
        
        // Shooting - Python: can_shoot and mouse_pressed[0] and fired == False and (time - last_shot)
        // Fix: Add proper shot cooldown check (Python code has bug - missing > shot_cooldown)
        let bullet = null;
        if (canShoot && pointer.isDown && !this.fired && (time - this.lastShot) >= CONFIG.SHOT_COOLDOWN) {
            bullet = angleDeg;
            this.fired = true;
            this.lastShot = time;
        }
        // Python: if mouse_pressed[0] == False: self.fired = False
        if (!pointer.isDown) {
            this.fired = false;
        }
        
        return { bullet, angleDeg };
    }
    
    drawLaserSight(angleDeg) {
        if (!this.scene.gameState.hasLaserSight) return;
        
        if (!this.laserLine) {
            this.laserLine = this.scene.add.graphics();
            // Python: Laser sight should be above player but below weapon
            // Player depth: 10, Weapon depth: 11, Laser depth: 10.5 (between them)
            this.laserLine.setDepth(10.5);
        }
        
        this.laserLine.clear();
        
        // Python: Calculate direction from player to mouse
        // Python: dx = mouse_x - player.rect.centerx, dy = mouse_y - player.rect.centery
        // Python: distance = sqrt(dx*dx + dy*dy), nx = dx/distance, ny = dy/distance
        // Python uses normalized direction vector, not angle
        // angleDeg is from atan2(xDist, yDist), which gives direction
        // For laser, we need to use the same direction as bullet flight (angleDeg)
        // But Python uses normalized vector, so we convert angleDeg to direction vector
        const angleRad = Phaser.Math.DegToRad(angleDeg);
        // Python: atan2(x_dist, y_dist) means: dx direction = sin(angle), dy direction = cos(angle)
        // But in Python, y increases downward, so we need to account for that
        // Python's atan2(x_dist, y_dist) gives angle where:
        // - x_dist positive (right) = angle positive
        // - y_dist positive (down) = angle positive
        // So direction vector: nx = sin(angle), ny = cos(angle) (for screen coordinates)
        const nx = Math.sin(angleRad);
        const ny = Math.cos(angleRad);
        
        // Python: Start point (gun position) - player.rect.centerx + nx * 30
        const startX = this.playerSprite.x + nx * 30;
        const startY = this.playerSprite.y + ny * 30;
        
        // Python: End point (extend to edge) - start_x + nx * 800
        const dist = 800;
        const endX = startX + nx * dist;
        const endY = startY + ny * dist;
        
        // Draw laser line with glow effect - Python: multiple layers
        // Outer glow
        this.laserLine.lineStyle(8, 0xff0000, 0.12); // (255, 0, 0, 30)
        this.laserLine.lineBetween(startX, startY, endX, endY);
        this.laserLine.lineStyle(4, 0xff0000, 0.24); // (255, 0, 0, 60)
        this.laserLine.lineBetween(startX, startY, endX, endY);
        // Core line
        this.laserLine.lineStyle(2, 0xff3232, 0.59); // (255, 50, 50, 150)
        this.laserLine.lineBetween(startX, startY, endX, endY);
    }
    
    clearLaserSight() {
        if (this.laserLine) {
            this.laserLine.clear();
        }
    }
}
